bin/faucet/src/main.rs

```diff
@@ -229,3 +196,133 @@ fn long_version() -> LongVersion {
         debug: option_env!("VERGEN_CARGO_DEBUG").unwrap_or_default(),
     }
 }
+
+fn create_new_router(faucet_state: FaucetState) -> Router {
+    Router::new()
+        .route("/", get(get_index_html))
+        .route("/index.js", get(get_index_js))
+        .route("/index.css", get(get_index_css))
+        .route("/background.png", get(get_background))
+        .route("/favicon.ico", get(get_favicon))
+        .route("/get_metadata", get(get_metadata))
+        .route("/get_tokens", post(get_tokens))
+        .layer(
+            ServiceBuilder::new()
+                .layer(TraceLayer::new_for_http())
+                .layer(SetResponseHeaderLayer::if_not_present(
+                    http::header::CACHE_CONTROL,
+                    HeaderValue::from_static("no-cache"),
+                ))
+                .layer(
+                    CorsLayer::new()
+                        .allow_origin(tower_http::cors::Any)
+                        .allow_methods(tower_http::cors::Any),
+                ),
+        )
+        .with_state(faucet_state)
+}
+
+async fn serve_faucet(config: FaucetConfig) -> anyhow::Result<()> {
+    let faucet_state = FaucetState::new(config.clone()).await?;
+
+    info!(target: COMPONENT, %config, "Initializing server");
+
+    let app = create_new_router(faucet_state);
+
+    let socket_addr =
+        config
+            .endpoint
+            .socket_addrs(|| None)?
+            .into_iter()
+            .next()
+            .ok_or(anyhow::anyhow!(
+                "Couldn't get any socket addrs for endpoint: {}",
+                config.endpoint
+            ))?;
+    let listener = TcpListener::bind(socket_addr).await.context("failed to bind TCP listener")?;
+
+    info!(target: COMPONENT, endpoint = %config.endpoint, "Server started");
+
+    axum::serve(listener, app).await.unwrap();
+    Ok(())
+}
+
+#[cfg(test)]
+mod test {
+    use std::{process::Command, str::FromStr, time::Duration};
+
+    use fantoccini::ClientBuilder;
+    use serde_json::{json, Map};
+    use tokio::time::sleep;
+    use url::Url;
+
+    use crate::{config::FaucetConfig, serve_faucet, stub_rpc_api::serve_stub};
+
+    #[tokio::test]
+    async fn test_website() {
+        let stub_node_url = Url::from_str("http://localhost:50051").unwrap();
+
+        // Start the stub node
+        let stub_url_clone = stub_node_url.clone();
+        tokio::spawn(async move { serve_stub(&stub_url_clone).await.unwrap() });
```

Comment:
This avoids the awkward `_clone` name at least.
```suggestion
        tokio::spawn({
            let stub_node_url = stub_node_url.clone();
            async move { serve_stub(&stub_url_clone).await.unwrap() }
        });
```